### PR TITLE
Use the Python logging module

### PR DESCRIPTION
**Pull request by erikos**
This will give us a better control about
the log levels for individual logs.

https://phabricator.endlessm.com/T21742

### CODE REVIEW

File: eoscompanion/main.py [hunk 1]
```diff
@@ -177,4 +177,7 @@ def main(args=None):
                 os.pathsep.join([
                     GLib.getenv('XDG_DATA_DIRS') or ''
                 ] + flatpak_export_share_dirs), True)
+
+    logging.basicConfig(level=logging.DEBUG)
```
Comment (smspillaz) [inline eoscompanion/main.py]: What is logging configured as during the tests? (this only affects the main application).

**Comment by smspillaz** [comment]
I think this will still require a configuration option in `/etc/eos-companion-app/config.ini` ... or did you intend to do that in a follow up PR?

File: Makefile.am.service.inc [hunk 2]
```diff
@@ -41,7 +41,6 @@ app_PYTHON += \
 	eoscompanion/responses.py \
 	eoscompanion/routes.py \
 	eoscompanion/service.py \
-	eoscompanion/util.py \
```
Comment (smspillaz) [inline Makefile.am.service.inc]: I don't understand why this change is necessary?

File: eoscompanion/main.py [hunk 3]
```diff
@@ -157,6 +157,31 @@ def do_activate(self):  # pylint: disable=arguments-differ
         logging.info('Activated')
         return Gio.Application.do_activate(self)
 
+CONFIG_FILE = '/etc/eos-companion-app/config.ini'
+CONFIG_SECTION = 'Companion App'
+CONFIG_NAME = 'loglevel'
+DEFAULT_LOG_LEVEL = logging.INFO
+
+def getLogLevel():
```
Comment (smspillaz) [inline eoscompanion/main.py]: Function naming style should be PEP8 complaint here (`get_log_level`). Also, please add a docstring.

File: eoscompanion/main.py [hunk 4]
```diff
@@ -157,6 +157,31 @@ def do_activate(self):  # pylint: disable=arguments-differ
         logging.info('Activated')
         return Gio.Application.do_activate(self)
 
+CONFIG_FILE = '/etc/eos-companion-app/config.ini'
```
Comment (smspillaz) [inline eoscompanion/main.py]: This path should not be hardcodeded. Have a look at https://github.com/endlessm/eos-companion-app-integration/blob/master/configuration-manager/main.c#L64 for details on which configuration file to load.

File: eoscompanion/main.py [hunk 5]
```diff
@@ -157,6 +157,31 @@ def do_activate(self):  # pylint: disable=arguments-differ
...
+
+def getLogLevel():
+    keyfile = GLib.KeyFile()
+    try:
+        keyfile.load_from_file(CONFIG_FILE, GLib.KeyFileFlags.NONE)
+    except GLib.Error:
+        return DEFAULT_LOG_LEVEL
+
+    try:
+        configLevel = keyfile.get_value(CONFIG_SECTION, CONFIG_NAME)
```
Comment (smspillaz) [inline eoscompanion/main.py]: Consider using `g_key_file_get_integer` which will throw `G_KEY_FILE_ERROR_INVALID_VALUE` in the event of a malformed config.

File: eoscompanion/main.py [hunk 6]
```diff
@@ -157,6 +157,31 @@ def do_activate(self):  # pylint: disable=arguments-differ
...
+    except GLib.Error:
```
Comment (smspillaz) [inline eoscompanion/main.py]: It is probably worth using `except GLib.Error as error: ... if error.matches` to filter for the errors that we will treat as non-fatal. I think those would be `G_KEY_FILE_ERROR_KEY_NOT_FOUND` `G_KEY_FILE_ERROR_GROUP_NOT_FOUND` and `G_KEY_FILE_ERROR_INVALID_VALUE`

File: Makefile.am.service.inc (hunk 2 above)
Comment (erikos) [reply Makefile.am.service.inc]: As discussed in the standup, as I remove the util.py, it only contained the custom logger.

File: eoscompanion/main.py (hunk 3 above)
Comment (erikos) [reply eoscompanion/main.py]: Fixed

File: eoscompanion/routes.py [hunk 7]
```diff
@@ -825,12 +815,11 @@ def _on_got_wrapped_stream(error, result):
...
-        'applicationId={applicationId}, contentId={contentId}'.format(
-            contentId=query['contentId'],
-            applicationId=query['applicationId'],
-            clientId=query['deviceUUID']
-        )
+        'Get content stream: clientId=%s, '
+        'applicationId=%s, contentId=%s',
+            query['deviceUUID'],
+            query['applicationId'],
+            query['contentId']
```
Comment (smspillaz) [inline eoscompanion/routes.py]: Indentation here looks a bit odd, did pylint insist on this?

File: eoscompanion/main.py (hunk 6 above)
Comment (smspillaz) [reply eoscompanion/main.py]: Still need to match the error here

File: eoscompanion/routes.py [hunk 8]
```diff
@@ -742,8 +741,7 @@ def _on_got_wrapped_stream(error, result):
             '''
             if error != None:
                 logging.warning(
-                    'Stream wrapping failed {error}'.format(error=error),
-                    file=sys.stderr
+                    'Stream wrapping failed {error}'.format(error=error)
```
Comment (smspillaz) [inline eoscompanion/routes.py]: Question: Where does `logging`'s output go? Does it go to the stdout or stderr? If there's a switch for it to go to the stderr my preference would be to turn that on.

File: eoscompanion/main.py [hunk 9]
```diff
@@ -157,7 +157,7 @@ def do_activate(self):  # pylint: disable=arguments-differ
         logging.info('Activated')
         return Gio.Application.do_activate(self)
 
-CONFIG_FILE = '/etc/eos-companion-app/config.ini'
+CONFIG_FILE = '/var/run/host/etc/eos-companion-app/config.ini'
```
Comment (smspillaz) [inline eoscompanion/main.py]: Good that this works in the sandbox - though I think this still needs to traverse the config directories in priority order. You may also need to add `--filesystem=/var/lib/eos-companion-app`

**Comment by smspillaz** [comment]
Okay, I squashed the branch a little and implemented the changes to read all three config files.

File: eoscompanion/main.py [hunk 10]
```diff
@@ -149,17 +149,79 @@ def do_dbus_register(self, connection, object_path):  # pylint: disable=argument
...
 
     def do_activate(self):  # pylint: disable=arguments-differ
         '''Invoked when the application is activated.'''
-        log('Activated')
+        logging.info('Activated')
         return Gio.Application.do_activate(self)
 
 
+COMPANION_APP_CONFIG_FILES = [
+    '/var/run/host/usr/share/eos-companion-app/config.ini',
```
Comment (erikos) [inline eoscompanion/main.py]: Is the order correct here? I have adjusted my config file in "/etc/eos-companion-app/config.ini" but "/usr/share/eos-companion-app/config.ini" takes precedence. That file owned by root is not writeable for me - while the config under /etc is for me as it has been created by the companion-app-helper.

If we keep this order maybe, we should check as well if the file has the 'loglevel' set an not bail out when the file is available.

Comment (smspillaz) [reply eoscompanion/main.py]: You're right - the order should go the other way around

Comment (erikos) [reply eoscompanion/main.py]: Fixed.

File: eoscompanion/main.py [hunk 11]
```diff
@@ -149,17 +149,79 @@ def do_dbus_register(self, connection, object_path):  # pylint: disable=argument
...
         '''Invoked when the application is activated.'''
-        log('Activated')
+        logging.info('Activated')
         return Gio.Application.do_activate(self)
 
 
+COMPANION_APP_CONFIG_FILES = [
+    '/var/run/host/etc/eos-companion-app/config.ini',
+    '/var/run/host/usr/share/eos-companion-app/config.ini',
+    # FIXME: /var currently does not bind in /run/host in the sandbox
```
Comment (cosimoc) [inline eoscompanion/main.py]: I'm pretty sure that you can explicitly mount `/var/lib/eos-companion-app` from the manifest...
Also, the `/var` path should take priority over the `/usr` one so it should be moved one line up

File: com.endlessm.CompanionAppService.json.in [hunk 12]
```diff
@@ -16,6 +16,7 @@
         "--filesystem=~/.config/dconf:ro",
         "--filesystem=/var/lib/flatpak:ro",
         "--filesystem=/var/endless-extra/flatpak:ro",
+        "--filesystem=host:ro",
```
Comment (cosimoc) [inline com.endlessm.CompanionAppService.json.in]: Would be nice if we could only mount `/usr/share/eos-companion-app` and `/etc/eos-companion-app` instead of the whole host FS... but no worries if that does not work.

File: eoscompanion/main.py [hunk 13]
```diff
@@ -149,17 +149,79 @@ def do_dbus_register(self, connection, object_path):  # pylint: disable=argument
...
 
     def do_activate(self):  # pylint: disable=arguments-differ
         '''Invoked when the application is activated.'''
-        log('Activated')
+        logging.info('Activated')
         return Gio.Application.do_activate(self)
 
 
+COMPANION_APP_CONFIG_FILES = [
+    '/var/run/host/etc/eos-companion-app/config.ini',
```
Comment (cosimoc) [inline eoscompanion/main.py]: `/var/run` is just a symlink to `/run`, so we should use the latter

File: eoscompanion/main.py [hunk 14]
```diff
@@ -149,17 +149,79 @@ def do_dbus_register(self, connection, object_path):  # pylint: disable=argument
...
+    config file. If we can not get the value we return the
+    default log level.
+    '''
+    keyfile = find_best_matching_config_file()
+
+    if not keyfile:
+        return DEFAULT_LOG_LEVEL
+
+    try:
+        config_level = keyfile.get_value(COMPANION_APP_CONFIG_SECTION,
```
Comment (cosimoc) [inline eoscompanion/main.py]: Should we not use one of the "typed" getters such as `keyfile.get_integer()` or `keyfiile.get_string()` here instead? It's not actually clear to me what the configuration file format is. The default configuration file that gets installed in `/usr` should also be changed in this patchset to include the new default.

File: eoscompanion/routes.py [hunk 15]
```diff
@@ -647,11 +641,8 @@ def on_splice_finished(src, result):
                     src.splice_finish(result)
                 except GLib.Error as splice_error:
                     # Can't really do much here except log server side
-                    log(
-                        'Splice operation on file failed: {error}'.format(
-                            error=splice_error
-                        ),
-                        file=sys.stderr
+                    logging.warning(
```
Comment (cosimoc) [inline eoscompanion/routes.py]: Why not `logging.error()`?

File: eoscompanion/routes.py [hunk 16]
```diff
@@ -725,9 +716,8 @@ def _on_got_wrapped_stream(error, result):
                 a tuple of an error or a stream and length.
                 '''
                 if error is not None:
-                    log(
-                        'Stream wrapping failed {error}'.format(error=error),
-                        file=sys.stderr
+                    logging.warning(
```
Comment (cosimoc) [inline eoscompanion/routes.py]: Why not `logging.error()`

File: eoscompanion/main.py (hunk 14 above)
Comment (smspillaz) [reply eoscompanion/main.py]: `keyfile.get_string()` is what we want here.

I think it is okay to leave it there without the option - the falling back to the hardcoded default is fine.

File: eoscompanion/routes.py (hunk 15 above)
Comment (smspillaz) [reply eoscompanion/routes.py]: Not sure if this is an error in the sense of the program state being inconsistent - this usually happens because the client hung up.

File: com.endlessm.CompanionAppService.json.in (hunk 12 above)
Comment (smspillaz) [reply com.endlessm.CompanionAppService.json.in]: Yeah, I am afraid that is not possible. Should file an upstream flatpak ticket.

File: eoscompanion/main.py (hunk 13 above)
Comment (erikos) [reply eoscompanion/main.py]: Fixed to /run...

**Comment by smspillaz** [comment]
Okay, per @erikos on slack, lets merge this now so that we can rebase other work on top of it.
